### PR TITLE
Add notification when report w/ segment has no data, but segment is unprocessed

### LINKED ISSUE #12255: When a custom segment is used in an API request and browser_archiving_disabled_enforce=1, throw an exception explaining to create the segment first
## Use case

As an API user, when I call the API to fetch a segmented report, and use a custom segment (ie. i didn't create it via segment editor but manually constructed the segment parameter in my API request), if the config.ini.php has  `browser_archiving_disabled_enforce=1`, my response will always show "no data", and I don't know why. I need to see clearly in the API response somehow that I need to manually create the segment in the segment editor first, before being able to fetch the data via the API for this segment. 

## Reproduce

* `browser_archiving_disabled_enforce=1` in config.ini.php
* call the API with a custom segment (ie. a segment which is not created in the Segment Editor)

Result: always show 0 / no data. 

## Steps

It's important to fix this issue because it is setup this way for some [Piwik service](https://www.innocraft.cloud/) customers.

### When `browser_archiving_disabled_enforce=1`  and the API is used with a `&segment=` parameter

1. When the segment exists in the DB and is set as `segmented reports are processed in real time (default)`, then the segment data will never be displayed in standard reports (it will however work in Real time reports). In case the data would never be displayed, we would need to throw an exception:

> The Segment is set to "segmented reports are processed in real time (default)" but Piwik has been configured to not process segmented reports in API requests. Therefore to see data for this report in the future, you need to edit your segment and choose "segmented reports are pre-processed (faster, requires cron)". Once you have edited your segment, it will take up to a few hours for your segments data to appear here.

2. When the segment does not exist in the DB (ie. a "custom" segment), then the data will never be displayed in standard reports (it will however work in real time reports). We would need to throw an exception:

> The Segment you specified has not been created yet in the Segment Editor and therefore the report data has not been pre-processed yet. To solve this, you need to go to Piwik and create this segment manually in the Segment editor (alternatively, you can create a new segment pragmatically using the `SegmentEditor.add` API). Once you have created the segment in the editor (or via API), this error message won't be displayed any more and you will see your segmented reports data within a few hours (once the segment data has been pre-processed). 
> Please note that you can already test whether your segment will work and match your users by using the `Live.getLastVisitsDetails` API. When using the  `Live.getLastVisitsDetails` API, you will see in real time which visits and actions were matched by your segment which can help you confirm your `&segment=` parameter is constructed well and works as expected.

Feedback welcome, the goal is to make it crystal clear to users why there is no data and what to do to get data.

### LINKED ISSUE #12256: When a segment is set to "reports are pre-processed" and viewing a report which hasn't yet been processed, display a notification
## Use case 

As a user when I view a report where a segment is applied, in some cases I will see "There is no data for this report" and may think that "My segment doesn't work!". But in many cases it is just a matter of waiting for some time and the segment data will be shown later once it's been processed. So I'd like to see a note that the data will be available later and I should check back after some time, rather than worry the service does not work. 

This is the use case on the [Piwik Cloud-hosted](https://www.innocraft.cloud/) for some customers.

## Solution

Whenever the requested segment hasn't been processed yet by `core:archive` for the requested date/period/site, we need to detect that "This segment hasn't yet been processed but will be later". 

* When we detect that the Segment exists in the DB and set to "reports are pre-processed", then we can display a message:

> Segment 'XYZ' not applied
> You are requesting data for the Segment 'XYZ'. Your segmented reports haven't yet been processed by the system and they will be available in a few hours. 
> Because all your reports for segment 'XYZ' will be processed by the system over the next few hours, we kindly invite you to check again in a few hours.
> Please note that you can already test whether your segment will work and match your users by applying your segment [a]in the Visitor Log[/a]. When using the Visitor Log, you can see in real time which visits and actions were matched by your segment. This can help you confirm your Segment is constructed well and matches your users and actions as you expect to. 

* We would link to the Visitor Log for the same date/period/site so user can see directly whether there will be later data in the report originally requested.
* This Modal shouldn't be displayed in pages where there are only Real time reports (visitor log, real time map, real time forms, real time videos, etc.)
* This is very similar to  #12253 and likely we would want to use a same UX solution & similar wording (the modal solution seems fine, or would it be better to use a Notice or Warning notification?)

Refs:

* Show an error in API response for "real time segments" and "custom segments" when `browser_archiving_disabled_enforce=1` in #12255
* Show a modal for "real time segments" when `browser_archiving_disabled_enforce=1` in #12253

### PR DESCRIPTION
**Pull request by diosmosis**
**Note: this PR has API changes.**

image

This PR makes two changes to the way notifications are handled/displayed:

* Transient notifications are now removed when the reporting page changes.
* ViewDataTables have a view property called 'notifications' which can be used to add new notifications when loading a report. This is not in the Config so is not API. (The actual creation of notifications is done in the widget loader directive.)

Individual changes:

* Remove 'notification_parser.js' and replace w/ `notification.service.js` angular service.
* Add 'notifications' property to ViewDataTable's View & handle in _dataTable.twig. The widget loader directive initiates the "parse" of the divs outputted in _dataTable.twig. (Also done by widget loading logic in dashboard code.)
* In reporting page controller, clear transient notifications after page load.
* Added `getSegmentByDefinition()` function to SegmentEditor `Model`.
* Add new event `Visualization.onLoadingError` that is triggered when there an exception is caught when loading report data.
* Add new event `Archive.noArchivedData` that is triggered when during an API request no archive data can be found.
* Use new event to display a notification when a report has no data and is for an unprocessed segment.

Fixes #12256 
Fixes #12255

### CODE REVIEW

File: plugins/SegmentEditor/templates/_unprocessedSegmentMessage.twig
Comment (sgiehl) [inline plugins/SegmentEditor/templates/_unprocessedSegmentMessage.twig]: Any reason for not adding this translatable?

Comment (diosmosis) [reply plugins/SegmentEditor/templates/_unprocessedSegmentMessage.twig]: Oh, right, forgot

**Comment by diosmosis** [comment]
Adding to 3.5.0 milestone in case it can go in, if not can remove it.

**Comment by diosmosis** [comment]
@tsteur would you have time to review this PR?

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [inline plugins/SegmentEditor/SegmentEditor.php]: do we need to use `getRawSegment()`?

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [inline plugins/SegmentEditor/SegmentEditor.php]: Not sure if it helps but there is also a `Piwik\Segment::willBeArchived()`

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [inline plugins/SegmentEditor/SegmentEditor.php]: What if there are multiple reports on the page and eg the first fails because of unprocessed and the next one because of some other reason? Then the notification wouldn't be sent?

File: plugins/CoreHome/templates/_dataTable.twig
Comment (tsteur) [inline plugins/CoreHome/templates/_dataTable.twig]: when multiple reports are rendered on one page and they all trigger a notification, will they all be added multiple times?

File: plugins/SegmentEditor/SegmentEditor.php
Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: That was just to be safe, but I guess it doesn't really help, I'll remove it.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [inline plugins/SegmentEditor/SegmentEditor.php]: Can there be a case where the segment used to exist, but not anymore, but we did archive this segment when it existed a few weeks ago but because there was really no match for this segment that time and the report therefore actually had no data, we would trigger this exception?

File: plugins/CoreHome/templates/_dataTable.twig
Comment (diosmosis) [reply plugins/CoreHome/templates/_dataTable.twig]: Depends on whether they use an ID or not. With an ID that doesn't change based on report, only one shows up. If no IDs are used, then I think multiple ones show up.

In this PR, since it uses an ID for the notification, there's only ever one.

File: core/Archive.php
Comment (tsteur) [inline core/Archive.php]: I feel like this might not be the best place for some reason. Eg it would stop archiving everything else afterwards? Eg I just recently implemented some logic to trigger an archiving of a segment within another archiving process and this segment might not be defined etc. Also there might be no ArchiveIds when there is simply no data etc? 

I feel like it might be more of an event that should be handled eg in the visualization if there is no data for a report, then check if it a segment was used and check if it will be archived or not.

The challenge be how to do this for API requests and that's why you likely added this event here. The report rendered doesn't seem like the best place either. The post processor may work for this event but not really sure either. It seems like there is no good place. 

I might have to think about this a bit longer but feeling like throwing exceptions here can cause lots of random side effects that we maybe aren't prepared for. I would love if it could be just handled in the visualization itself, and then somehow also for API requests (when the root request is an API request maybe only, not when executing an API request within a regular UI request or so)

Comment (tsteur) [reply core/Archive.php]: Maybe it would already help to do it only if eg `Root::isRootApiRequest() && !$ArchiveTrigger` or so.

Comment (diosmosis) [reply core/Archive.php]: Yes I originally had it in the Visualization, and it worked for one ticket, but not for the other one. I thought about doing it on Request.dispatch for API methods but that has odd edge cases as well (eg, we'd have to check if what we got is an empty datatable, but what about datatable\maps? and what about plugins that return something else? or plugins that just do something that doesn't trigger archiving?).

Checking for root API request that isn't in cron archiving sounds better to me. This would be in the event handler, correct? Or did you mean something else?

Comment (tsteur) [reply core/Archive.php]: Yeah I totally see what you mean. I'll have another think and get back to you later (in about 1.5 hours) or at the latest tmrw morning.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [inline plugins/SegmentEditor/SegmentEditor.php]: I wonder if we also need to try `in_array(urlencode($segment), $segmentsToPreprocess)` and/or `in_array(urldecode($segment), $segmentsToPreprocess)` as I know we accept it somehow encoded and not encoded or so.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [inline plugins/SegmentEditor/SegmentEditor.php]: Maybe this method could be made reusable similar to `Segment::willBeArchived` and / or maybe the logic can or has to be merged?

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [reply plugins/SegmentEditor/SegmentEditor.php]: I'm not quite sure if that actually works but it might do. When archiving is disabled, then it might be archived during the cron... and only need to take a while to become available. But also only if the date is recent... I think we need to take many more things into consideration eg when the date is like 2 months back, it likely won't be archived?

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [reply plugins/SegmentEditor/SegmentEditor.php]: It might be similar to `Rules::isSegmentPreProcessed()`

Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: I think `Segment` will urldecode, so if it's in the config/DB encoded I suppose somehow it could work. I can add the check for that.

File: core/Archive.php
Comment (tsteur) [reply core/Archive.php]: I can't think of any other place. Maybe we could tag the events with `@ignore` so they won't be public API allowing us to change things later? Then in the event we would only throw the exception if possible when we are in an actual API request (`Request::isRootApiRequest()` or similar). 

Also we need to check maybe if there it really hasn't been archived or whether there was just no data to be archived? Not sure but something to think about.

I'm thinking we actually need another method `Request::isRootApiRequestHandlingMethod($method . '.' . $action)` as we maybe don't want to throw an exception if an API request eg without a segment executes another API request with a segment (using `Request::processRequest`)? It gets quite complicated though if you know what I mean. I basically just want to avoid throwing an exception maybe when it is not "expected" to not really break any possible logic etc. But we could maybe try without this at first (but then we never add the check unless we do run into a problem).

File: core/Plugin/Visualization.php
Comment (tsteur) [inline core/Plugin/Visualization.php]: Maybe we can tag this event with `@ignore` as well?

File: core/Plugin/Visualization.php
Comment (tsteur) [inline core/Plugin/Visualization.php]: Actually I wonder if we would need some UI tests for each visualization when there is no data to see if any errors occur and whether all visualisations still render nicely? Eg if some view template is expecting a dataTable to be set when it possibly isn't it may fail... not sure if that is easy doable though and might be bit "painful" to write a test for.

File: core/Archive.php
Comment (tsteur) [reply core/Archive.php]: > Haven't looked at isRootApiRequest(), but if another request is executed inside one request, then the "sub-request" isn't the root one right?

Right, but so far we are not handling this in that method.

Comment (tsteur) [reply core/Archive.php]: Or lets say that method only checks if originally an API request is executed but not whether we are currently in that request or not. This would likely me a different method.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: Re: `willBeArchived()` that will return true if the segment is going to be processed in the next cron (for some date), but here we need to know if the segment should have been processed already, not if it's going to be processed.

> I think we need to take many more things into consideration eg when the date is like 2 months back, it likely won't be archived?

Didn't think of that... In this case we really do need some way of knowing what's next to be archived in the CronArchive run... Would it be good enough to check if the period in the request includes the current day and if the current site has had new visits? (using RawLogDao::hasSiteVisitsBetweenTimeframe() as is done in CronArchive)

Comment (tsteur) [reply plugins/SegmentEditor/SegmentEditor.php]: Wouldn't we want to show a message like "This report data will become available soon"? or something similar? All good though if this is not covered in the PR.

Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: I'm not sure I understand what you mean, there is already a message in the notification that says "Data for this segment will become available in a few hours". Is that what you're expecting?

File: plugins/SegmentEditor/SegmentEditor.php
Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: Do you see potential for reuse w/ this logic? I'm not sure where it would go if made re-usable.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: In case it's still not clear why I can't use willBeArchived(), is that it won't help w/ custom segments, which will also display "no data", but will never be archived in this case.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (diosmosis) [reply plugins/SegmentEditor/SegmentEditor.php]: Performed a test where I:

* Enabled browser archiving & loaded a preprocessed segment on a day w/ no data.
* Verified that there was an archive w/ the segment hash.
* Disabled browser archiving & loaded the same reports (w/ same day & segment).
* Saw that the notification was not displayed.
* Went to a more recent day (after dropping the archive tables) & saw that the notification was displayed.

So this shouldn't be an issue.

File: plugins/SegmentEditor/SegmentEditor.php
Comment (tsteur) [reply plugins/SegmentEditor/SegmentEditor.php]: OK I got confused with `willBeArchived` is not helpful so I thought we wouldn't show that message. 

> "Data for this segment will become available in a few hours". Is that what you're expecting?

Yes that's what I was expecting. 

> In case it's still not clear why I can't use willBeArchived(), is that it won't help w/ custom segments, which will also display "no data", but will never be archived in this case.

Still not quite sure if I'm getting it but I reckon there are those 2 cases:
* the data is currently not there but it will become available
* the data is currently not there and it won't become available.

**Comment by diosmosis** [comment]
After discussing it appears to risky to introduce this change in 3.5.1, will remove this + the related issues from the current milestone.

CC @tsteur / @mattab

**Comment by mattab** [comment]
> If we only show this notification for dates that are < 2 months old, will it still be useful?
> For context: we don't really know if an older date will be archived by the next cron:archive run, so if we say data for every date will become available, it's possible some users will just wait forever & then get angry.

I think we are able to know whether a given date in the past will be archived. we wouldn't want to pick a certain date (eg. 2 months) because that depends on the settings of how far log data was purged. In theory (without reading the whole thread) the historical data will be processed for a segment, as long as we still have the raw data (maybe `haveLogsBeenPurged` is useful there) @diosmosis

**Comment by diosmosis** [comment]
@mattab So what about `min(2 months, $logDataPurgeTime)`? so if the log data purges older than 1 month, we check for 1 month, if it's 3 months, we check for 2 months?

CC @tsteur since he's more aware of the limitations of knowing what core:archive will do than I am and can probably explain the issue better.

**Comment by tsteur** [comment]
>  In theory (without reading the whole thread) the historical data will be processed for a segment, as long as we still have the raw data

I'm not quite sure since it might depend on when the archiving ran the last time? See https://github.com/matomo-org/matomo/blob/3.5.1/core/CronArchive.php#L1556-L1577 

I reckon it may be best to start a document or comment to write all the things done we know influence when something will be archived and when not. For example the `process_new_segments_from` setting etc.

**Comment by mattab** [comment]
> I'm not quite sure since it might depend on when the archiving ran the last time? See https://github.com/matomo-org/matomo/blob/3.5.1/core/CronArchive.php#L1556-L1577

The historical segments should always be processed within a day or two... because whenever we next call "month" archiving and then "year" archiving, it will pre-process the last ~ 700 days (last 2 years) / last 24 months / etc. 

(assuming we have a compatible process_new_segments_from value eg. beginning_of_time)

**Comment by mattab** [comment]
> So what about min(2 months, $logDataPurgeTime)? so if the log data purges older than 1 month, we check for 1 month, if it's 3 months, we check for 2 months?

@diosmosis don't understand the "2 months", AFAIK we don't want to use a fixed date as most of the time it will even pre-process 1 year, or 2 years, or 3 years of historical segment data.

**Comment by diosmosis** [comment]
That was @tsteur’s suggestion on slack, I was deferring to his knowledge of the archiving process.

**Comment by diosmosis** [comment]
I think @tsteur had some bigger issues w/ this one specifically regarding the inability to determine when a segment will be archived in the future. Probably needs more discussion w/ @tsteur & @mattab.

File: core/API/Request.php
Comment (tsteur) [inline core/API/Request.php]: I'm not quite sure if this works / makes too much sense? `$_GET/$_POST['method']` will be overwritten during API calls within the code etc. And it seems to be only called once but without a parameter? And what is the difference to `isRootRequestApiRequest()` when called without a parameter or so? I've tried to understand what is happening here but couldn't.

File: core/API/Request.php
Comment (tsteur) [inline core/API/Request.php]: I wonder if we need to clarify API request vs root API request... in this case here we would be also returning a method when an API request is being executed as part of a regular request... maybe we also need some wordings for it to better talk about it? not sure though.

File: CHANGELOG.md
Comment (tsteur) [inline CHANGELOG.md]: Can you remove the events from changelog again as we "ignore" them and don't consider them API?

**Comment by tsteur** [comment]
Had a look through the code again but to really say whether it covers all cases or not, what else we need to consider etc I would need a day or two to understand things fully. As long as nothing is API and we can change things easily later I presume it should be fine to merge as long as we don't say "Data will become available later" when it actually maybe won't.

**Comment by diosmosis** [comment]
Applied some changes, but nothing w/ regard to when the message is displayed or not.

Maybe the following changes would help:
- Only show the message for a recent period (where recent could be within 2 months or some other date).
- Change the message to display "Data for this Segment will become available in a few hours when processing completes. If it does not, there may be a problem."

What do you think @mattab / @tsteur ?

File: plugins/SegmentEditor/templates/_unprocessedSegmentMessage.twig
Comment (mattab) [inline plugins/SegmentEditor/templates/_unprocessedSegmentMessage.twig]: Wondering about the `|raw` here, maybe it could lead to XSS for specially crafted segments, ie. maybe `segmentName` needs to be escaped?

File: plugins/SegmentEditor/UnprocessedSegmentException.php
Comment (mattab) [inline plugins/SegmentEditor/UnprocessedSegmentException.php]: Header comment missing

File: plugins/SegmentEditor/lang/en.json
Comment (mattab) [inline plugins/SegmentEditor/lang/en.json]: maybe this could be renamed to `CustomUnprocessedSegmentNoData` as there was no other key called this?

**Review by mattab — CHANGES_REQUESTED** [review]
Feedback

*  please add tests for the API outputs for the 3 use cases covered here: 

1. custom segment with archiving disabled -> won't be ever available
2. segment created in editor as real-time processed but web archiving is disabled so they need to set the segment to pre-processed
3. segment created in editor as pre-processed, will be available soon

tests will ensure we don't regress this key information/new user feedback.

**Comment by mattab** [comment]
@diosmosis 

> Maybe the following changes would help:
>   Only show the message for a recent period (where recent could be within 2 months or some other date).
>    Change the message to display "Data for this Segment will become available in a few hours when processing completes. If it does not, there may be a problem."

Change which message to this one?
 :+1: otherwise for the proposed message.  but we need to display this message only when we know logs are still available for this date/period and haven't been purged yet (according to the Data retention configured in the Privacy settings in Matomo admin). Also check the message wouldn't be displayed on date/periods before the site creation date.

**Comment by diosmosis** [comment]
> Change which message to this one?

Both, since they both tell the user when data will become available.

**Comment by diosmosis** [comment]
Found issues w/ this one, don't merge yet.

**Comment by diosmosis** [comment]
should be good now but could use more testing I think

**Comment by diosmosis** [comment]
Issues should be fixed

**Comment by mattab** [comment]
@diosmosis Build is failing at least in https://travis-ci.org/matomo-org/matomo/jobs/411881882

**Comment by diosmosis** [comment]
Build should be passing (UI test failures are unrelated to this PR, will have to be fixed in another PR)

**Comment by MatomoForumNotifications** [comment]
This pull request has been mentioned on **Matomo forums**. There might be relevant details there:

https://forum.matomo.org/t/custom-segment-archiving-question/49322/3

**Comment by metalocator** [comment]
Hi,

Really wishing there was a "force" option only on the API side.

We can't allow customers to create segments via the web because they're often too data intensive, risking an outage with a complex segment with millions of records.

However, we have in-app dashboards that are generated dynamically based on API calls, and to create every dashboard segment beforehand wouldn't be practical.  We have dashboards that can report on a per profile basis, and one customer might have 500,000 profiles.  Also then if they delete profiles, we need to keep the segment list in sync, not to mention we would be required to maintain thousands of segments many of which would never be used.

**Comment by sgiehl** [comment]
@metalocator If you have a feature request or suggestion for improvement please consider open a new issue, if non exist yet.
Comments on older and closed pull requests or issues, will not get any attention when considering future improvements...

**Pull request merged**
